apps/omg/lib/omg/ethereum_event_listener/measure.ex

```diff
@@ -0,0 +1,69 @@
+# Copyright 2019 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.EthereumEventListener.Measure do
+  @moduledoc """
+  Counting business metrics sent to DataDog
+  """
+
+  import OMG.Status.Metric.Event, only: [name: 2]
+
+  alias OMG.Status.Metric.Datadog
+  alias OMG.Status.Metric.Tracer
+
+  @supported_events [
+    [:process, OMG.EthereumEventListener],
+    [:process, OMG.EthereumEventListener.Core],
+    [:trace, OMG.EthereumEventListener],
+    [:trace, OMG.EthereumEventListener.Core]
+  ]
+  def supported_events, do: @supported_events
+
+  def handle_event([:process, OMG.EthereumEventListener.Core], %{events: events}, state, _config) do
+    :ok = Datadog.gauge(name(state.service_name, :events), length(events))
+  end
+
+  def handle_event([:process, OMG.EthereumEventListener], %{}, state, _config) do
+    value =
+      self()
+      |> Process.info(:message_queue_len)
+      |> elem(1)
+
+    :ok = Datadog.gauge(name(state.service_name, :message_queue_len), value)
+  end
+
+  def handle_event([:trace, _], %{}, state, _config) do
+    # TODO change to compiler flags once we're running releases
+    is_child_chain =
+      Enum.any?(Application.started_applications(), fn
+        {:omg_child_chain, _, _} -> true
+        {_, _version, _desc} -> false
+      end)
+
+    if is_child_chain do
```

Comment:
This bit is necessary even with the two different tracers? (`OMG.ChildChainRPC.Tracer` and `OMG.Watcher.Tracer`)

Reply:
Ha, yes, I've spend quite a bit of time here. You can specify a tracer by setting an argument to a decorator to a `span` or a `trace`.
`@decorate span(tracer: MyApp.OtherTracer)`
`@decorate trace(tracer: MyApp.OtherTracer)`
Macros in general get evaluated at compilation and `ethereum_event_listener` is being used by both Watcher and Child Chain. If we were using releases, we could do a pre-compile check that would allow us to figure out if we're doing a Watcher compilation or a Child Chain compilation by doing something like:
```
if Code.ensure_loaded?(OMG.Watcher) do
@decorate span(tracer: OMG.Watcher.Tracer)
else
@decorate span(tracer: OMG.ChildChain.Tracer)
end
```
and this would produce a Macro wrapper that injects code depending on whether a module is present at compilation. 

Currently, because we use the `launcher.py` script and `mix run` to launch the app this isn't possible, because erlang's VM always loads ALL modules into the VM.